code/modules/mob/dead/observer/freelook/marker/marker_net.dm

```diff
@@ -0,0 +1,225 @@
+// MARKER NET
+//
+// The datum containing all the chunks.
+/datum/markernet
+	/// Name to show for VV and stat()
+	var/name = "Camera Net"
+	/// The chunks of the map, mapping the areas that the cameras can see.
+	var/list/chunks
+	/// The image cloned by all chunk static images put onto turfs cameras cant see
+	var/image/alpha_mask
+	/// An assoc list of all vision source. vision_source = vision_type
+	var/list/visionSources
+	/// List of all eyes in our markernet, used when creating a new chunk
+	var/list/eyes
+
+/datum/markernet/New()
+	chunks = list()
+	visionSources = list()
+	eyes = list()
+	alpha_mask = image('icons/hud/alpha_mask.dmi', null, "alpha_mask")
+	alpha_mask.plane = OBSCURITY_MASKING_PLANE
+	alpha_mask.appearance_flags = RESET_TRANSFORM|RESET_ALPHA|RESET_COLOR|KEEP_APART
+
+// Used only in one place, if you want it to use somewhere else - do safety checks first. e.g. chunkGenerated()
+/datum/markernet/proc/generateChunk(x, y, z)
+	x &= ~(CHUNK_SIZE - 1)
+	y &= ~(CHUNK_SIZE - 1)
+	return chunks["[x],[y],[z]"] = new /datum/markerchunk(x, y, z, visionSources, alpha_mask)
+
+/// Checks if a chunk has been Generated in x, y, z.
+/datum/markernet/proc/chunkGenerated(x, y, z)
+	x &= ~(CHUNK_SIZE - 1)
+	y &= ~(CHUNK_SIZE - 1)
+	return chunks["[x],[y],[z]"]
+
+/// Updates what the aiEye can see. It is recommended you use this when the aiEye moves or it's location is set.
+/datum/markernet/proc/visibility(mob/dead/observer/signal/eye)
+	var/list/visibleChunks = list()
+	if(eye.loc)
+		var/static_range = eye.static_visibility_range
+		var/x1 = max(0, eye.x - static_range) & ~(CHUNK_SIZE - 1)
+		var/y1 = max(0, eye.y - static_range) & ~(CHUNK_SIZE - 1)
+		var/x2 = min(world.maxx, eye.x + static_range) & ~(CHUNK_SIZE - 1)
+		var/y2 = min(world.maxy, eye.y + static_range) & ~(CHUNK_SIZE - 1)
+
+		for(var/x = x1 to x2 step CHUNK_SIZE)
+			for(var/y = y1 to y2 step CHUNK_SIZE)
+				var/datum/markerchunk/chunk = chunkGenerated(x, y, eye.z)
+				if(chunk)
+					visibleChunks |= chunk
+
+	var/list/remove = eye.visibleChunks - visibleChunks
+	var/list/add = visibleChunks - eye.visibleChunks
+
+	for(var/datum/markerchunk/chunk as anything in remove)
+		chunk.remove(eye)
+
+	for(var/datum/markerchunk/chunk as anything in add)
+		chunk.add(eye)
```

Comment:
Might be worth making `remove` and `add` accept lists so you don't need so much overhead on the proc calls.